File: 0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
    var stack []rune
    var closeBracket = map[rune]rune{
        '(' : ')',
        '{' : '}',
        '[' : ']',
    }
    for _, char := range s {
        // if the char is open bracket, push into stack
        if char == '(' || char == '{' || char == '[' {
            stack = append(stack, char)
        } else {
            // if the char is close bracket and the stack is empty, return false
            if len(stack) == 0 {
                return false
            }
            // assume the stack is not empty,
            // if the char is not close bracket of top the stack, return false, otherwise, pop from stack 
            lastIndex := len(stack)-1
            topStack := stack[lastIndex]
            if  char != closeBracket[topStack] {
                return false
            } else {                
                stack = stack[:len(stack)-1]            
            }
        }
    }
    return len(stack) == 0
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        // Iterate each character on s
        // If it is open bracket, push it into a stack,
        // Else pop an open braket from stak, check if the close bracket match

        stack<char> openbs; // should contains open brackets

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' ||
                s[i] == '{' ||
                s[i] == '[') {
                    openbs.push(s[i]);
                    continue;
                }
            
            if (openbs.empty()) {
                return false;
            }
            
            char openb = openbs.top();

            switch (openb) {
                case '(':
                    if (s[i] != ')') {
                        return false;
                    }
                    break;
                case '{':
                    if (s[i] != '}') {
                        return false;
                    }
                    break;
                case '[':
                    if (s[i] != ']') {
                        return false;
                    }
                    break;
            }

            openbs.pop();
        }

        return openbs.empty();
    }
};